src/portals/Screenshot.cpp

```diff
@@ -0,0 +1,123 @@
+#include "Screenshot.hpp"
+#include "../core/PortalManager.hpp"
+#include "../helpers/Log.hpp"
+#include "../helpers/MiscFunctions.hpp"
+#include <regex>
+
+CScreenshotPortal::CScreenshotPortal() {
+    m_pObject = sdbus::createObject(*g_pPortalManager->getConnection(), OBJECT_PATH);
+
+    m_pObject->registerMethod(INTERFACE_NAME, "Screenshot", "ossa{sv}", "ua{sv}", [&](sdbus::MethodCall c) { onScreenshot(c); });
+    m_pObject->registerMethod(INTERFACE_NAME, "PickColor", "ossa{sv}", "ua{sv}", [&](sdbus::MethodCall c) { onPickColor(c); });
+
+    m_pObject->registerProperty(INTERFACE_NAME, "version", "u", [](sdbus::PropertyGetReply& reply) -> void { reply << (uint32_t)(2); });
+
+    m_pObject->finishRegistration();
+
+    Debug::log(LOG, "[screenshot] init successful");
+}
+
+void CScreenshotPortal::onScreenshot(sdbus::MethodCall& call) {
+    sdbus::ObjectPath requestHandle;
+    call >> requestHandle;
+
+    std::string appID;
+    call >> appID;
+
+    std::string parentWindow;
+    call >> parentWindow;
+
+    std::unordered_map<std::string, sdbus::Variant> options;
+    call >> options;
+
+    Debug::log(LOG, "[screenshot] New screenshot request:");
+    Debug::log(LOG, "[screenshot]  | {}", requestHandle.c_str());
+    Debug::log(LOG, "[screenshot]  | appid: {}", appID);
+
+    bool isInteractive = options.count("interactive") && options["interactive"].get<bool>();
+
+    // make screenshot
+    const std::string FILE_PATH = "/tmp/xdph_screenshot.png";
+    const std::string SNAP_CMD  = "grim " + FILE_PATH;
+    const std::string SNAP_INTERACTIVE_CMD = "grim -g \"$(slurp)\" " + FILE_PATH;
+
+    std::unordered_map<std::string, sdbus::Variant> results;
+    results["uri"] = "file://" + FILE_PATH;
+
+    remove(FILE_PATH.c_str());
+
+    if (isInteractive) {
+        execAndGet(SNAP_INTERACTIVE_CMD.c_str());
+    } else {
+        execAndGet(SNAP_CMD.c_str());
+    }
+
+    uint32_t responseCode = fileExists(FILE_PATH) ? 0 : 1;
+
+    auto reply = call.createReply();
+    reply << responseCode;
+    reply << results;
+    reply.send();
+}
+
+void CScreenshotPortal::onPickColor(sdbus::MethodCall& call) {
+    sdbus::ObjectPath requestHandle;
+    call >> requestHandle;
+
+    std::string appID;
+    call >> appID;
+
+    std::string parentWindow;
+    call >> parentWindow;
+
+    Debug::log(LOG, "[screenshot] New PickColor request:");
+    Debug::log(LOG, "[screenshot]  | {}", requestHandle.c_str());
+    Debug::log(LOG, "[screenshot]  | appid: {}", appID);
+
+    const std::string PICK_COLOR_CMD = "grim -g \"$(slurp -p)\" -t ppm -";
```

Comment:
why don't we use our own hyprpicker, if the user has it installed, and only fall back to slurp?